src/engine/terrain.cpp

```diff
@@ -363,3 +289,167 @@ void Terrain::Draw() {
+
+          current_tile->shading[0] = vertices[(tile_y * 5) + tile_x].lighting;
+          current_tile->shading[1] = vertices[(tile_y * 5) + tile_x + 1].lighting;
+          current_tile->shading[2] = vertices[((tile_y + 1) * 5) + tile_x].lighting;
+          current_tile->shading[3] = vertices[((tile_y + 1) * 5) + tile_x + 1].lighting;
+        }
+      }
+    }
+  }
+
+  std::fclose(fh);
+
+  Update();
+}
+
+void Terrain::LoadHeightmap(const std::string& path, int multiplier) {
+  std::string p = u_find(std::string(path).c_str());
+
+  PLImage image;
+  if (!plLoadImage(p.c_str(), &image)) {
+    LogWarn("Failed to load the specified heightmap, \"%s\" (%s)!\n", p.c_str(), plGetError());
+    return;
+  }
+
+  if (image.width < 65 || image.height < 65) {
+    plFreeImage(&image);
+    LogWarn("Invalid image size for heightmap, %dx%d vs 65x65!\n", image.width, image.height);
+    return;
+  }
+
+  // Each channel is encoded with specific data
+  // red = height
+  // green = texture
+
+  unsigned int chan_length = image.width * image.height;
+
+  auto* rchan = static_cast<float*>(u_alloc(chan_length, sizeof(float), true));
+  memset(rchan, 0, chan_length);
+  uint8_t* pixel = image.data[0];
+  for (unsigned int i = 0; i < chan_length; ++i) {
+    rchan[i] = static_cast<int>(*pixel) * multiplier; //(static_cast<int>(*pixel) - 127) * 256;
+    pixel += 4;
+  }
+
+  auto* gchan = static_cast<uint8_t*>(u_alloc(chan_length, sizeof(uint8_t), true));
+  memset(gchan, 0, chan_length);
+  pixel = image.data[0] + 1;
+  for (unsigned int i = 0; i < chan_length; ++i) {
+    gchan[i] = *pixel;
+    pixel += 4;
+  }
+
+  plFreeImage(&image);
+
+  for (unsigned int chunk_y = 0; chunk_y < TERRAIN_CHUNK_ROW; ++chunk_y) {
+    for (unsigned int chunk_x = 0; chunk_x < TERRAIN_CHUNK_ROW; ++chunk_x) {
+      Chunk& current_chunk = chunks_[chunk_x + chunk_y * TERRAIN_CHUNK_ROW];
+      for (unsigned int tile_y = 0; tile_y < TERRAIN_CHUNK_ROW_TILES; ++tile_y) {
+        for (unsigned int tile_x = 0; tile_x < TERRAIN_CHUNK_ROW_TILES; ++tile_x) {
+          Tile* current_tile = &current_chunk.tiles[tile_x + tile_y * TERRAIN_CHUNK_ROW_TILES];
+          unsigned int aaa = (chunk_y * 4 * 65) + (chunk_x * 4);
+          current_tile->height[0] = rchan[aaa + (tile_y * 65) + tile_x];
+          current_tile->height[1] = rchan[aaa + (tile_y * 65) + tile_x + 1];
+          current_tile->height[2] = rchan[aaa + ((tile_y + 1) * 65) + tile_x];
+          current_tile->height[3] = rchan[aaa + ((tile_y + 1) * 65) + tile_x + 1];
+
+          current_tile->texture = gchan[aaa + (tile_y * 65) + tile_x];
+
+          // hrm...
+          current_tile->shading[0] =
+          current_tile->shading[1] =
+          current_tile->shading[2] =
+          current_tile->shading[3] = 255;
+        }
+      }
+
+      // Find the maximum and minimum points
+      for (auto& tile : current_chunk.tiles) {
+        for (float i : tile.height) {
+          if (i > max_height_) {
```

Comment:
`max_height_` should be initialised to either its minimum possible value or one of the tiles before this loop I think.

Reply:
fixed